name: Test gevent py2.7

on:
  push:
    branches:
      - master
      - release/**

  pull_request:

# Cancel in progress workflows on pull_requests.
# https://docs.github.com/en/actions/using-jobs/using-concurrency#example-using-a-fallback-value
concurrency:
  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
  cancel-in-progress: true

permissions:
  contents: read

env:
  BUILD_CACHE_KEY: ${{ github.sha }}
  CACHED_BUILD_PATHS: |
    ${{ github.workspace }}/dist-serverless

jobs:
  test:
    name: gevent, python 2.7, ubuntu-20.04
    runs-on: ubuntu-20.04
    container: python:2.7
    timeout-minutes: 30

    steps:
      - uses: actions/checkout@v3

      - name: Setup Test Env
        run: |
          pip install coverage "tox>=3,<4"

      - name: Test gevent
        uses: nick-fields/retry@v2
        with:
          timeout_minutes: 15
          max_attempts: 2
          retry_wait_seconds: 5
          shell: bash
          command: |
            set -x # print commands that are executed
            coverage erase

            # Run tests
            ./scripts/runtox.sh "py2.7-gevent" --cov=tests --cov=sentry_sdk --cov-report= --cov-branch &&
            coverage combine .coverage* &&
            coverage xml -i

      - uses: codecov/codecov-action@v3
        with:
          token: ${{ secrets.CODECOV_TOKEN }}
          files: coverage.xml

  check_required_tests:
    name: All gevent tests passed or skipped
    needs: test
    # Always run this, even if a dependent job failed
    if: always()
    runs-on: ubuntu-20.04
    steps:
      - name: Check for failures
        if: contains(needs.test.result, 'failure')
        run: |
          echo "One of the dependent jobs have failed. You may need to re-run it." && exit 1
